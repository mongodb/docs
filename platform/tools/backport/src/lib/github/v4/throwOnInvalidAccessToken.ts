import { isEmpty, difference } from 'lodash';
import { maybe } from '../../../utils/maybe';
import { BackportError } from '../../BackportError';
import { getGlobalConfigPath } from '../../env';
import { GithubV4Exception } from './apiRequestV4';

export function throwOnInvalidAccessToken({
  error,
  repoOwner,
  repoName,
  globalConfigFile,
}: {
  error: GithubV4Exception<unknown>;
  repoOwner: string;
  repoName: string;
  globalConfigFile?: string;
}) {
  function getSSOAuthUrl(ssoHeader?: string) {
    const matches = ssoHeader?.match(/url=(.*)/);
    if (matches) {
      return matches[1];
    }
  }

  const statusCode = error.githubResponse.status;

  switch (statusCode) {
    case 200: {
      const repoNotFound = error.githubResponse.data.errors?.some(
        (error) =>
          error.type === 'NOT_FOUND' && error.path?.join('.') === 'repository',
      );

      const grantedScopes =
        error.githubResponse.headers['x-oauth-scopes'] || '';
      const requiredScopes =
        error.githubResponse.headers['x-accepted-oauth-scopes'] || '';
      const ssoHeader = maybe(error.githubResponse.headers['x-github-sso']);

      if (repoNotFound) {
        const hasRequiredScopes = isEmpty(
          difference(requiredScopes.split(','), grantedScopes.split(',')),
        );

        // user does not have permission to the repo
        if (!hasRequiredScopes) {
          throw new BackportError(
            `You do not have access to the repository "${repoOwner}/${repoName}". Please make sure your access token has the required scopes.\n\nRequired scopes: ${requiredScopes}\nAccess token scopes: ${grantedScopes}`,
          );
        }

        // repo does not exist
        throw new BackportError(
          `The repository "${repoOwner}/${repoName}" doesn't exist`,
        );
      }

      const repoAccessForbidden = error.githubResponse.data.errors?.some(
        (error) => error.type === 'FORBIDDEN',
      );

      const ssoAuthUrl = getSSOAuthUrl(ssoHeader);

      // user does not have permissions
      if (repoAccessForbidden && ssoAuthUrl) {
        throw new BackportError(
          `Please follow the link to authorize your personal access token with SSO:\n\n${ssoAuthUrl}`,
        );
      }
      break;
    }

    case 401:
      throw new BackportError(
        `Please check your access token and make sure it is valid.\nConfig: ${getGlobalConfigPath(
          globalConfigFile,
        )}`,
      );

    default:
      return undefined;
  }
}
