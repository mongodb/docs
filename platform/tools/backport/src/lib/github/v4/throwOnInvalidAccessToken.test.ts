import { throwOnInvalidAccessToken } from './throwOnInvalidAccessToken';

describe('throwOnInvalidAccessToken', () => {
  describe('when status code is', () => {
    it('should handle invalid access token', () => {
      const error = {
        githubResponse: {
          status: 401,
          headers: {},
        },
      } as any;

      return expect(() =>
        throwOnInvalidAccessToken({
          repoOwner: 'elastic',
          repoName: 'kibana',
          error,
        }),
      ).toThrow('Please check your access token and make sure it is valid');
    });

    it('should handle SSO error', () => {
      const error = {
        githubResponse: {
          status: 200,
          headers: { 'x-github-sso': 'required; url=https://ssourl.com' },
          data: {
            errors: [{ type: 'FORBIDDEN' }],
          },
        },
      } as any;

      return expect(() =>
        throwOnInvalidAccessToken({
          repoOwner: 'elastic',
          repoName: 'kibana',
          error,
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should handle non-existing repo', () => {
      const error = {
        githubResponse: {
          status: 200,
          headers: {
            'x-oauth-scopes': 'a,b,c',
            'x-accepted-oauth-scopes': 'a,b,c',
          },
          data: {
            errors: [{ type: 'NOT_FOUND', path: ['repository'] }],
          },
        },
      } as any;

      return expect(() =>
        throwOnInvalidAccessToken({
          repoOwner: 'elastic',
          repoName: 'kibana',
          error,
        }),
      ).toThrow(`The repository "elastic/kibana" doesn't exist`);
    });

    it('should handle insufficient permissions (oauth scopes)', () => {
      const error = {
        githubResponse: {
          status: 200,
          headers: {
            'x-oauth-scopes': 'a,b',
            'x-accepted-oauth-scopes': 'a,b,c',
          },
          data: {
            errors: [{ type: 'NOT_FOUND', path: ['repository'] }],
          },
        },
      } as any;

      return expect(() =>
        throwOnInvalidAccessToken({
          repoOwner: 'elastic',
          repoName: 'kibana',
          error,
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('should not handle unknown cases', () => {
      const error = {
        githubResponse: {
          status: 500,
          headers: {},
        },
      } as any;

      return expect(
        throwOnInvalidAccessToken({
          repoOwner: 'elastic',
          repoName: 'kibana',
          error,
        }),
      ).toBe(undefined);
    });
  });
});
