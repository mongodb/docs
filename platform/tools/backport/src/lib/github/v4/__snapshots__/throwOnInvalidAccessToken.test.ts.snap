// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throwOnInvalidAccessToken when status code is should handle SSO error 1`] = `
"Please follow the link to authorize your personal access token with SSO:

https://ssourl.com"
`;

exports[`throwOnInvalidAccessToken when status code is should handle insufficient permissions (oauth scopes) 1`] = `
"You do not have access to the repository "elastic/kibana". Please make sure your access token has the required scopes.

Required scopes: a,b,c
Access token scopes: a,b"
`;
